File: 2022/go/06.go
package main

import "fmt"

type Day6 struct{}

func (*Day6) Day() int {
	return 6
}

func (*Day6) IsTest() bool {
	return false
}

func (puzzle *Day6) Run(input string) {
	fmt.Printf("Part1: %d\n", puzzle.findStartOfPacket(input, 4))
	fmt.Printf("Part2: %d\n", puzzle.findStartOfPacket(input, 14))
}

func (puzzle *Day6) findStartOfPacket(input string, packetSize int) int {
	length := len(input) - packetSize

	for i := 0; i < length; i++ {
		slice := input[i : i+packetSize]
		if puzzle.isStartOfPacket(slice) {
			return i + packetSize
		}
	}

	return -1
}

func (puzzle *Day6) isStartOfPacket(packet string) bool {
	tester := map[rune]bool{}

	for _, c := range packet {
		if tester[c] {
			return false
		}
		tester[c] = true
	}

	return true
}

File: 2022/cs/06.cs
namespace Day6Namespace;

public class Day6 : IPuzzle
{
    public int Day => 6;
    public bool IsTest => false;

    public void Run(string input)
    {
        int part1 = FindStartOfPacket(input, 4);
        Console.WriteLine($"Part1: {part1}");

        int part2 = FindStartOfPacket(input, 14);
        Console.WriteLine($"Part2: {part2}");
    }

    private static int FindStartOfPacket(string input, int packetSize)
    {
        int length = input.Length - packetSize + 1;

        for (int i = 0; i < length; i++)
        {
            ReadOnlySpan<char> span = input.AsSpan(i, packetSize);
            if (IsStartOfPacket(span))
                return i + packetSize;
        }

        return -1;
    }

    private static readonly HashSet<char> tester = new();

    private static bool IsStartOfPacket(ReadOnlySpan<char> value)
    {
        tester.Clear();

        foreach (char c in value)
        {
            if (tester.Add(c) == false)
                return false;
        }

        return true;
    }
}
